### PR TITLE
Populating Python Package

### PR DESCRIPTION
**Pull request by mberaha**
In the Python package now we can
1. run the algorithms Neal2 and Neal8 for univariate normal models, using DP or PY mixings
2. estimate the density on a grid
3. estimate the clustering

In the DP & PY mixings, there are utilities for prior elicitation.
In the example/ folder, two notebooks collect examples of usage of the package

### CODE REVIEW

File: pybmix/core/chain.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        return out
+
+
+    def _get_extractor(self, param_name):
+        if self.chain is None:
+            def extractor(x): return get_field(self._deserialize(x), param_name)
+        else:
+            def extractor(x): return get_field(x, param_name)
+
+        return extractor
```
Comment (brunoguindani) [inline pybmix/core/chain.py:108]: Wouldn't it be better to just initialize the appropriate `extractor` in the constructor and store it as a class member? It seems pointless to pass it to all these member functions

File: pybmix/core/mixing.py [hunk 2]
```diff
@@ -112,13 +106,75 @@ def _check_args(self, total_mass, total_mass_prior):
...
+        self.prior_proto.fixed_values.discount = discount
+
+    def _check_args(self, strength, discount):
+        if strength <= 0:
+            raise ValueError(
+                "Parameter 'strength' must be strictly greater than zero")
+
+        if discount >= 1 or discount <= 0:
+            raise ValueError(
+                "Parameter 'discount' must be in the range (0, 1")
```
Comment (brunoguindani) [inline pybmix/core/mixing.py:180]: String is missing a )

File: pybmix/core/chain.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
+import logging
+import numpy as np
+from google.protobuf.pyext._message import RepeatedScalarContainer
+
+from pybmix.utils.proto_utils import get_field
+import pybmix.proto.matrix_pb2 as matrix_pb2
+
+
+class MmcmChain(object):
```
Comment (brunoguindani) [inline pybmix/core/chain.py:9]: Is it spelled "Mmcm" on purpose...? After the memoizer thingy I can't claim to know anything.
Also, if it is indeed misspelled, I think it looks nicer in all caps

**Review by brunoguindani — COMMENTED** [review]
Thank you, it's great to see that a functional package is slowly building up 😊

File: examples/Estimate Univariate Density.ipynb [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "y = np.concatenate(\n",
+    "    [np.random.normal(loc=3, size=100), np.random.normal(loc=-3, size=100)])\n",
+    "plt.hist(y)\n",
```
[the remark this answers is not in the document]
Comment (mberaha) [reply examples/Estimate Univariate Density.ipynb:32]: Sure thing, nice catch!

File: pybmix/core/chain.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        >>> chain = mixture_model.get_chain()
+        >>> card_chain = chain.extract("cluster_states[0].cardinality")
+        """
+        chain = self.chain if self.chain is not None else self.serialized_chain
+        extractor = self._get_extractor(param_name)
+        out = None
+
+        # we need to perform these checks because we're checking also
+        # for base classes
+        if isinstance(extractor(chain[0]), RepeatedScalarContainer):
```
[the remark this answers is not in the document]
Comment (mberaha) [reply pybmix/core/chain.py:55]: that's true, I will add a check

File: pybmix/core/chain.py (hunk 1 above)
Comment (mberaha) [reply pybmix/core/chain.py:108]: But the `extractor` depends on the `param_name`, and we might want to extract several parameters from the same chain.

**Review by brunoguindani — APPROVED** [review]

**Pull request merged**